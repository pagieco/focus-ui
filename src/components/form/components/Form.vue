<script>

export const Methods = {
  Get: 'get',
  Post: 'post',
  Action: 'action',
};

export const EncTypes = {
  urlencoded: 'application/x-www-form-urlencoded',
  multipart: 'multipart/form-data',
  plain: 'text/plain',
};

export default {
  name: 'focus-form',

  props: {
    /**
     * The method used to submit the form.
     */
    method: {
      type: String,
      default: Methods.Post,
      validator: val => Object
        .values(Methods)
        .includes(val),
    },

    /**
     * Where to send form-data on submittal.
     */
    action: {
      type: String,
    },

    /**
     * Space separated list of character encodings.
     */
    acceptCharset: {
      type: String,
    },

    /**
     * Grants the broswer the ability to autocomplete input elements.
     */
    autoComplete: {
      type: Boolean,
      default: false,
    },

    /**
     * The encoding type when submitting content to the server.
     */
    encType: {
      type: String,
      validator: val => Object
        .keys(EncTypes)
        .concat([null])
        .includes(val),
    },
  },

  computed: {
    formEnctype() {
      return EncTypes[this.encType];
    },

    formAutocomplete() {
      return this.autoComplete ? 'on' : 'off';
    },
  },
};

</script>

<template>
  <form :method="method"
        :action="action"
        :accept-charset="acceptCharset"
        :autocomplete="formAutocomplete"
        :enctype="formEnctype"
        v-on="$listeners">
    <slot />
  </form>
</template>
