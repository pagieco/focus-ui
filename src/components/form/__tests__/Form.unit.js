import { shallowMount } from '@vue/test-utils';
import Form, { EncTypes } from '../components/Form.vue';

const shallowWrap = (propsData = {}) => shallowMount(Form, { propsData });

describe('Form', () => {
  it('is a valid vue component', () => {
    expect(Form).toBeAComponent();
  });

  it('matches the snapshot', () => {
    expect(shallowMount(Form).element).toMatchSnapshot();
  });

  it('uses the method prop', () => {
    const wrapper = shallowWrap({ method: 'post' });

    expect(wrapper.attributes('method')).toBe('post');
  });

  it('validates on the correct method prop', () => {
    const prop = Form.props.method;

    expect(prop.validator).toBeInstanceOf(Function);
    expect(prop.validator('get')).toBeTruthy();
    expect(prop.validator('post')).toBeTruthy();
    expect(prop.validator('non-existing-method')).toBeFalsy();
  });

  it('uses the action prop', () => {
    const wrapper = shallowWrap({ action: 'post-url' });

    expect(wrapper.attributes('action')).toBe('post-url');
  });

  it('uses the accept-charset prop', () => {
    const wrapper = shallowWrap({ acceptCharset: 'utf-8' });

    expect(wrapper.attributes('accept-charset')).toBe('utf-8');
  });

  it('uses the auto-complete prop', () => {
    const wrapper = shallowWrap({ autoComplete: true });

    expect(wrapper.attributes('autocomplete')).toBe('on');
  });

  it('uses the enc-type prop', () => {
    const wrapper = shallowWrap({ encType: 'urlencoded' });

    expect(wrapper.attributes('enctype')).toBe(EncTypes.urlencoded);
  });

  it('validates on the correct enc-type prop', () => {
    const prop = Form.props.encType;

    expect(prop.validator).toBeInstanceOf(Function);
    expect(prop.validator('urlencoded')).toBeTruthy();
    expect(prop.validator('multipart')).toBeTruthy();
    expect(prop.validator('plain')).toBeTruthy();
    expect(prop.validator('something-exotic')).toBeFalsy();
  });
});
