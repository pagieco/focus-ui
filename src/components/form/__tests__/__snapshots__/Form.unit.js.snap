// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches the snapshot 1`] = `
<form
  autocomplete="off"
  method="post"
/>
`;
